### ISSUE TITLE
Online Portfolio Selection > Production-ready models

### ISSUE THREAD

**Issue by Eriz11** [hudson-and-thames/mlfinlab#408]
Hey **mlfinlab** team!

It's Eriz from the **Alpha Team at Darwinex**.

Huge congrats for the development of the `Online Portfolio Selection` module; the API is genuinely created together with the docs!

Just a little "enhancement" in a question form:

Let's say that I want to use the `RMR` method in production code. As the prices I'm feeding are the `close prices` of the `assets`, I guess (**if not, please correct me**) that the portfolio weights that are associated with each date are the ones to allocate for the following day (i.e. weights on `2019-05-26` in the final `all_weights` DataFrame are the weights that should be allocated on the following day's open > `2019-05-27`).

In this particular case, the allocation strategy needs to have and X quantity of observations based on the `window` parameter to get the following prediction. If after today's close I want to run this method to get tomorrow's open allocations what I believe is that I could:

1) Call `.allocate()` on the whole dataset plus the new data as the last row in the DataFrame of the assets close prices and then get the `weights` attribute (pretty inefficient but easy as some data providers actually just update the overall dataset when new observations are available).

2) Call `.allocate()` on just the last X `window` observations for the calculation to work (it seems much more practical, but is algorithm specific).

¿Is my understanding correct on how to apply this and the inner workings for production-like environments?

**PD**: The enhancement bit is to create a little example for the new comers in the docs when new data wants to be used for future prediction. Even to be more complete it would be awesome to adapt the different classes with a method that will throw to the back-end all the inner workings of new data prediction depending on the method (maybe I missed it and it is implemented?).

**PD2**: Using open prices if available to simulate the strategy better in this case adding them into another DataFrame and taking then for the returns calculations.

Cheers and again many thanks for all the work!

**Comment by PanPip** [comment]
Hey @Eriz11 

Thanks for the issue. Tagging the author @akwon31.

**Comment by akwon31** [comment]
Hi @Eriz11 

The weights displayed for ``2019-05-26`` in the final ``all_weights`` is actually the weights predicted from data until ``2019-05-25``. Therefore, if we had {A:0.25, B:0.25, C:0.5} for ``2019-05-26``, it actually means based on the data until ``2019-05-25`` we decided to choose those weights for the day's start. The final returns are calculated from the change in daily returns for each portfolio. Also calling ``.weights`` gives you the most recent prediction based on the entire data.

To your first observation of calling ``.allocate()`` on the whole dataset for the last price, it is necessary to do that if the particular algorithm looks back for the whole duration (i.e. CORN, OLMAR-2, etc).

To your second observation, that should be a lot faster, but it does depend on which one you are choosing. ``RMR`` in particular does have its specific window frame, so as long as you input the correct number of data points and initial weight, it should be fine. However, for the CORN strategies, it is based on the entire historical window, so that might not be the best choice.

For *PD*, I think what you've done in *1* and *2* are perfectly valid to recalculate and append the results, but for *PD2*, I'm slightly reluctant to do so as simulating can be in different forms especially with different transaction costs considerations (something that I have not incorporated).

Thanks for pointing out these things, I should clarify these issues in the docs.

**Comment by Eriz11** [comment]
Hey @akwon31 !

Many thanks for the rapid response!

> The weights displayed for 2019-05-26 in the final all_weights is actually the weights predicted from data until 2019-05-25. Therefore, if we had {A:0.25, B:0.25, C:0.5} for 2019-05-26, it actually means based on the data until 2019-05-25 we decided to choose those weights for the day's start. The final returns are calculated from the change in daily returns for each portfolio. Also calling .weights gives you the most recent prediction based on the entire data.

So, let's say that today `2020-06-26` after the close of X market I get all the `close` data up to and including today, input that into the `.allocate()` method (_without taking into account here that I could optimize the algorithm by just passing the `window` data for some implementations_) and the `.weights` attribute would be the weights that are chosen for `Monday start` in this case (i.e. `2020-06-29`).

If my assumption and understanding is correct (**correct me again if I'm wrong**), the implementation is even more brilliant and straightforward!

> For PD, I think what you've done in 1 and 2 are perfectly valid to recalculate and append the results, but for PD2, I'm slightly reluctant to do so as simulating can be in different forms especially with different transaction costs considerations (something that I have not incorporated).

Understood. I meant not as a "must" but more like an accesory implementation, but I'm sure that you will have better and more complete ideas for it!

Many thanks again!

**Comment by akwon31** [comment]
@Eriz11 
> So, let's say that today 2020-06-26 after the close of X market I get all the close data up to and including today, input that into the .allocate() method (without taking into account here that I could optimize the algorithm by just passing the window data for some implementations) and the .weights attribute would be the weights that are chosen for Monday start in this case (i.e. 2020-06-29).

This is correct.

**Comment by Jackal08** [comment]
Hi @Eriz11,

Just want to make double sure there is no confusion. You give it todays close with the expectation that you can execute at todays close, not the next days open. That would not be the correct way to apply the algorithm. 

In practice you pass through the best bids before the closing auction, then place your trades with the expectations that price wont be radically different in the auction.

**Comment by Eriz11** [comment]
Hey @Jackal08 ! 

Thanks for joining and for the clarification! 

Got it. Your notes are totally reasonable so that the execution prices don't diverge a lot from the ones just passed to the algorithm.

That even puts more emphasis in the need of the computation being efficient and fast, doesn't it?

**Comment by Jackal08** [comment]
Computation time should be reasonable but I dont see it as a blocker. 

The main hurdle I see is selecting a universe of stocks that have the properties that the algorithms try to exploit. 

A smaller universe of the correct stocks is much faster to compute and has highet PnL.

**Comment by andrewcztrack** [comment]
Hi all,

So the algorithm assumes the close and the open on the following day are the same?
Would there be a way to add slippage to make the backtests more reasonble?

Kind regards,
Andrew

**Comment by andrewcztrack** [comment]
I mean essentially you have weights which translates to an amount of share bought. So you have a percentage of your total capital divided by the current price of the shares? This would give you the amount of shares to buy in real terms?

**Comment by Eriz11** [comment]
Got it @Jackal08. 

I didn't mean computation to be a blocker but something to take care of if you want to have the least divergence between the supposed close price and the final execution price (and also depending on the final stock universe and production env).

In terms of analyzing the stocks that have the exploitable property, regime shift models together with other implementations could make a good job (some stuff coming via the DWX Alpha Team videos!).

Thanks for the discussion and inputs @mlfinlab team!

**Comment by Jackal08** [comment]
@Eriz11 absolutely agree. The actual execution price is a challenge for this algorithm and a very fast compute time would help to minimise this.

For some algos you dont need the entire history so you can just pass a small window fpr the computation. Can also move to cloud computing if one has the budget.

Very excited about the videos! Ping us if you have any questions. :)

**Comment by Jackal08** [comment]
Hi @andrewcztrack, it assumes that you can execute at the close. Thats all, nothing about the next day. Only that you can execute at the price quoted. :)

**Comment by andrewcztrack** [comment]
hi @Jackal08 !! thank you!
mm very interesting. Would you be able to give an outline of how the backtest might be able to be done.
so we get vector v of the weights calculated daily?

Kind regards and thanks,
Andrew

**Comment by andrewcztrack** [comment]
@akwon31 @Jackal08 @Eriz11 

Theorectilly how would implementing this model in production work.

For example, would the steps be -

- Data is loaded into a database.
- the algorithm is calculated, portfolio weights are calculated.
- the parameters are then optimised.
- new weights are calculated.
- stock values are calculated base on the weights
- orders are submitted on the start of the trading day?

Any comments would be much appreciated.

**Comment by Jackal08** [comment]
> @akwon31 @Jackal08 @Eriz11
> 
> Theorectilly how would implementing this model in production work.
> 
> For example, would the steps be -
> 
> * Data is loaded into a database.
> * the algorithm is calculated, portfolio weights are calculated.
> * the parameters are then optimised.
> * new weights are calculated.
> * stock values are calculated base on the weights
> * orders are submitted on the start of the trading day?
> 
> Any comments would be much appreciated.

Hi @andrewcztrack,

You would, just before the auction, pass through the prices to the algorithm, get weights, buy that portfolio in the auction or pick up left over liquidity.

Then you do the same thing the next rebalance period.

I want to stress this part: if you are buying at the next day open or the next day anything, you are doing it wrong.

**Comment by andrewcztrack** [comment]
Hi @Jackal08 !!

Thank you !
Question, there is hyperparamter tuning using optuna on the examples.
would not this need to be done before buy orders are executed.
Best,
Andrew

**Comment by Jackal08** [comment]
How to optimise a strategy is still an open question. My recommendation is to study the strategy, build a basket of instruments that allow you to exploit an inefficiency. Next do some hyper-parameter tuning to find a relatively stable set or use the universal portfolios. 

Your hope is that the stability holds in out-of-sample. If you re-optimise daily, I think you are going to cause a lot of problems.

**Comment by Eriz11** [comment]
Hey @Jackal08, @akwon31 and teammates, 

Now that we have started an interesting conversation about OLPS I have some questions that I believe will benefit the community. If you could share your valuable and experienced thoughts it would be awesome.

**1) Using alternative representations instead of time bars for returns calculation**. Could be the better showcased normality of those representations of use in the different strategies of the OLPS module?

**2) Edge look-up**. At the end of the day, all assets will have mean reverting, momemtum and pattern matching periods (not sure about the pattern matching one though); I guess that the first approach could be to analyze the assets in the portfolio and try to dissect whether some of them have more of one or another. What about a mixture of models?

**3) Timeframe for the training**. As you know, there is the common principle in data science about "more data is always better"; however, I'm not particularly agree with that statement (although in the case of portfolio management my believe is that ther could less dynamism once arranged). With the algorithms that use all the history to be trained, any suggestion on how to handle the training timeframe? 

**One last question**. I guess that all the algorithms out of the benchmarks (i.e. that get insight data), all the others run (whether or not they have a window) sample by sample without any look ahead of data, so that we could take the backtesting (although in sample) as something reliable. Just a confirmation check to scare the look-ahead bias ghosts :)

Best and thanks in advance!

**Comment by PanPip** [comment]
Tagging @Jackal08

**Comment by Jackal08** [comment]
Hi @Eriz11,

Thanks for your patience.

1.) I don't know how you will use dollar bars or the likes because each time series will have a different timestamp at which a sample was taken. I have heard that the ETF trick could be used to help, although I am not sure how. 

2.) Selecting a universe of assets that have the properties that you are trying to exploit in a given OLPS algo is very important. Some assets are more mean reverting or trending than others. Selection will be very important. A mixture of OLPS alogs into another overlaying strategy is interesting but now we are faced with the problem: If you have 1000s of strategies, or portfolios, or funds, you are still stuck with the asset allocation problem of just optimising a handful of stocks. 

3.) I agree, I also don't think that adding more data is the solution. The most relevant data and more observations is more helpful than just a crazy amount of features. Not sure how to answer this one, Alex Kwon can but at the moment he is not able to respond. Perhaps when he is done at his current post. 

4.) Hmm not sure what you mean here, would you mind to please rephrase?

Best
Jacques

**Comment by Eriz11** [comment]
Hey @Jackal08!

Thanks for the follow up!

1) Understood. It is true that I didn't consider the variable timestamp. Not very suitable then.

2) The thing here (from my humble perspective) is the common issue with historical data; depending on the look-back window you choose, you might see mean-reverting, trending or pattern matching situations. So it seems that is not "robust" enough to use it like that. However, the approach that I could consider most interesting could be to use regime switching models to analyze the actual situation of the specific assets, then use a meta-model architecture to use one type of algorithm or another. Does it seem right to you?

In the last weeks, I have created the Mendel Framework, an open sourced framework (it would be out very soon in the Darwinex GitHub) that will enable to manage DARWIN portfolios quantitatively. I have created a wrapper around the OLPS module for users to use the algorithms inside so maybe we could get some interesting insights while users do use it! 

3) Sure, will wait for those insights from @akwon31!

4) Let me rephrase it. My assumption is that the algorithms in the OLPS module do "back-test" with the event-driven approach, that is, when I get for example the final cummulative returns, the work done behind is totally realistic in terms of passing one observation at a time to the algorithm, computing the weights for that datetime (without having look-ahead bias or insight bias like it happens in the benchmarks algorithms in which you see the whole history first, then select for example the BestStock and show those returns) and doing that one at a time to raise a final cummulative return. Did I explain myself better this time? As said, just a confirmation check to scare the look-ahead bias ghosts :)

Many thanks again!

**Comment by sword134** [comment]
@Eriz11 Did you ever get clearance about your question number 4? I too would like to know if the results are replicable in a live trading environment to some degree.

**Comment by Eriz11** [comment]
Hi @sword134!

Not yet. It is indeed a very important point to understand if the data is handled ala event-driven to then be extrapolated to live environments. Hope to get an experienced answer soon!

**Comment by Jackal08** [comment]
Hi @Eriz11, 

2. That sounds like an interesting idea. I like the perspective of identifying regimes. The one concern based on my finance background is that certain stocks are defensive, have a low beta in the CAPM model where others will have a high beta. I think different stocks perform differently in different regimes. I don't think you get a regime that explains all movement. 

In a high vol regime based on the index, you will find that different sectors and industries behave differently. I think this is where a risk model like Barra can add a lot of value. Even the Alphalens from Quantopian, classic Fama French / smart beta stuff.

side) The Mendel Framework sounds interesting.

4. I have a bit more of a strict definition of an event-driven back tester, the OLPS uses a for loop backtester and makes some simplifying assumptions but they don't look into the future, I think you are safe from the look-ahead bias in this respect. The main challenge you will face is the assumption that we can trade at the price, passed to the OLPS algorithms, there will be slippage here.

---

@sword134 and @Eriz11, it's very important to understand that as a group, we are making zero promises as to the performance of these algorithms in real-life environments. You can be confident that it is sourced from the best journals, with authors from strong academic and institutional backgrounds. I have seen many funds implement the OLPS algos but I can't comment on their success. 

This part is left to the user and this is where I think a platform like Darwinex adds a lot of value. We have done all the academic and implementation heavy lifting, now its up to the user to extract alpha. :)

**Comment by sword134** [comment]
@Jackal08 excellent reply clears up alot of the confusion so to speak.

I have but one question, say one does find an OLPS that works rather well and is not a k-top algo. Then how does one filter out the small positions?

Say you have 5 stocks A, B, C, D and E.
With a k-top you can tell the algo to only trade 3 of them at a time, this solves the problem. However certain algo's like the exponential gradient does not have a k-top variant, so how does one filter out the the small position sizes that can seem insignificant at first glance. It obviously would have to be filtered before the algo calculates the return etc so as one does not simply do it post production.

**Comment by andrewcztrack** [comment]
Hi @Jackal08 ! really interesting!
What happened to the research notebooks?

**Comment by sword134** [comment]
@andrewcztrack they locked the notebooks and documentation behind a paywall (patreon). It also seems now that they are trying to dodge issues/questions on github in favour of redirecting them to their slack channel (which you also need to pay for).

The thing is, no person who comes across this library is going to start paying them to get access to the docs, notebooks or even normal help with bugs and questions if they don't see how the product works first. In order to be able to test the product you need to have access to the things that are now locked behind a paywall.

**Comment by Jackal08** [comment]
Hi @andrewcztrack, 

We have recently moved towards an open-core model and have put the additional value add items like the documentation and notebooks under a paid-for service. Everything is available via [Patreon](https://www.patreon.com/HudsonThames)

@sword134, we hope that you come to see the huge amount of work and effort the team has put behind these implementations. If you find that the docs or notebooks is a value add, we would love it if you purchased them. Alternatively, the literature is available and we invite you to implement you own library from first principals and support it for free.

**Comment by sword134** [comment]
@Jackal08 I understand your reasoning for making it a paid-for service but please consider how repulsive it seems for people who are just now finding your library. They have limited tutorials available to them and little documentation, why do you think that they will pay for it? I am personally in that boat having just found the library and still struggling to see its application in a trading manner (OLPS) due to an issue I posted here on github. Replying with "visit our slack" isn't convincing me to buy the paid-for service since there isn't even a guarantee that my problem can be resolves there.
Besides slack really isn't efficient for making issues and what not, its harder to see historical issues and their answers so one might end up making a duplicate question.

**Comment by Jackal08** [comment]
@sword134 you're taking a very strong stance. My team has done a lot for promoting financial machine learning. We are not stopping you from achieving your goals, all the material you need to be a success is available to you, and I think everyone will agree that if anything, we are enabling people. 

If you want to build the implementations of OLPS from the textbooks yourself, you are entirely right to do so. It's a bit of a cheek though that you're criticising me for not wanting to do so for free.

FYI lots of people are paying for our service because they like it - because its a value add. No one is forcing you. If you don't want to pay for it, that's ok - but please don't be upset with us if we don't want to work for free.

**Comment by sword134** [comment]
@Jackal08 Im not saying it should be free, I am saying that perhaps you could do trial system or similar, I'm just saying that the majority of people might be repulsed by a paid model. I do support your project and I do also think I will become a donor in the future if all goes well. I'm just citing how I think the majority of people are going to act once finding the repo.

**Comment by Jackal08** [comment]
@sword134, but it is free, all the code is free... Just not the documentation and notebooks. 

We have done all the heavy lifting, all you have to do is use the code and try to extract alpha. We don't provide trading strategies, we don't make any promises in that respect. What we do is: build the very best algorithms that the literature has to offer. It's now your responsibility to take it to the next step.

Re your comment on stagnation, we are on a record-breaking month and its only day 3 of this new change. We made the right decision. Now we are going to reinvest that money into building an even better product so that we can service the needs of our clients, even better.

**Comment by Jackal08** [comment]
Hi @Eriz11,

I am going to close this ticket due to the debate that went on. Please do let me know if you have questions and we will do our best to help. Really love the work your team is doing at Darwinex.

**Comment by andrewcztrack** [comment]
Hi @Jackal08  for the record the service that your team has done for financial machine learning research has been outstanding. I am in your debt greatly for the remarkable work your team has and i sincerely appreciate, all the way from sydney australia!
I had a look at the patreon page and i will joing a tier once i have a clear view of a project.
Just to understand can i still download the updated library from pypi. If you dont mind me asking, what are some the future plans for the library.
Kind regards and thanks,
Andrew

**Comment by Jackal08** [comment]
Hi @andrewcztrack, 

Thank you very much for your kind words. I will share it with the team and I am sure it will brighten their day. 

At the moment, all of the code is still available on PyPi via `pip install mlfinlab`. We are still deciding exactly how the open-core model will look. Rest assured that we are committed to helping the ML in Fin community grow. I can't make any promises but our first step is to monetize the documentation and notebooks, find a balance as to what should be free and paid for and if we don't see a really good revenue stream, then we will need to make a plan.

Best regards
Jacques
